src/contradictory_claims/models/sbert_models.py

```diff
@@ -0,0 +1,382 @@
+    warmup_steps = math.ceil(len(train_dataloader_embed) * epochs / batch_size * 0.1)  # 10% of train data for warm-up
+
+    # now to train the final layer
+    train_dataset = ClassifierDataset(df_train, tokenizer=tokenizer)
+    val_dataset = ClassifierDataset(df_val, tokenizer=tokenizer)
+    if enable_class_weights==False:
+        class_weights=None
+    else:
+        class_weights = train_dataset.class_weights()
+
+    train_dataloader = DataLoader(dataset=train_dataset,
+                                  batch_size=batch_size,
+                                  collate_fn=collate_fn,
+                                  shuffle=True)
+    val_dataloader = DataLoader(dataset=val_dataset,
+                                batch_size=1,
+                                collate_fn=collate_fn)
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    criterion = nn.CrossEntropyLoss(weight=class_weights.to(device))
+    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
+    model.to(device)
+    accuracy_stats = {"train": [],
+                      "val": [],
+                      }
+    loss_stats = {"train": [],
+                  "val": [],
+                  }
+
+    print("------TRAINING STARTS----------")  # noqa: T001
+    ## train embedding layer
+    unfreeze_layer(model.embedding_model)
+    model.embedding_model.fit(train_objectives=[(train_dataloader_embed, train_loss_embed)],
+                              evaluator=evaluator,
+                              epochs=1,
+                              evaluation_steps=1000,
+                              warmup_steps=warmup_steps,
+                              )  # train the Transformer layer
+    freeze_layer(model.embedding_model)
+
+
+    for e in range(epochs):
+        train_epoch_loss = 0
+        train_epoch_acc = 0
+        model.train()
+
+        for sentence1, sentence2, label in tqdm(train_dataloader):
+            label = label.to(device)
+            optimizer.zero_grad()
+            y_train_pred = model(sentence1, sentence2)
+
+            train_loss = criterion(y_train_pred, label)
+            train_acc = multi_acc(y_train_pred, label)
+
+            train_loss.backward()
+            optimizer.step()
+
+            train_epoch_loss += train_loss.item()
+            train_epoch_acc += train_acc.item()
+
+        # VALIDATION
+        with torch.no_grad():
+
+            val_epoch_loss = 0
+            val_epoch_acc = 0
+
+            model.eval()
+            for sentence1, sentence2, label in val_dataloader:
+                label = label.to(device)
+                y_val_pred = model(sentence1, sentence2)
+
+                val_loss = criterion(y_val_pred, label)
+                val_acc = multi_acc(y_val_pred, label)
+
+                val_epoch_loss += val_loss.item()
+                val_epoch_acc += val_acc.item()
+
+        loss_stats['train'].append(train_epoch_loss / len(train_dataloader))
```

Comment:
Perhaps we should be using the callbacks from the call to _eval_during_training() in model.fit() ?

Reply:
hmm, I get an idea of what you are proposing, 
let me find out a solution closest to what you're proposing, I haven't exactly seen callbacks in torch